### PR TITLE
Uninterpreted Functions in Specifications

### PR DESCRIPTION
**Pull request by ranjitjhala**
See 

- [uif00](flux-tests/tests/pos/surface/uif00.rs)
- [uif01](flux-tests/tests/pos/surface/uif01.rs)

and the `neg` variants for example.

Several things could be improved in due course, IMO

* Using plain `Symbol` or `String` for UIF-name?
* Gross `UFArg` introduced to format args with parens if needed

### CODE REVIEW

**Review by nilehmann — APPROVED** [review]
@ranjitjhala looks good to me. I think it's ok to use `Symbol` for now, but I think eventually what we need is to have an analogous to `DefId` and use that in the lower levels.

File: flux-driver/src/callbacks.rs [hunk 1]
```diff
@@ -194,10 +203,17 @@ impl<'a, 'genv, 'tcx> CrateChecker<'a, 'genv, 'tcx> {
...
+                            Ok(fn_sig) => {
+                                let fn_sig =
+                                    desugar::desugar_fn_sig(genv, &adt_sorts, def_id, fn_sig)?;
+                                Wf::new(genv).check_fn_sig(&fn_sig)?;
+                                genv.register_fn_sig(def_id.to_def_id(), fn_sig);
+                            }
+                            Err(span) => {
+                                genv.sess.emit_err(errors::InvalidAliasApplication { span });
+                            }
+                        }
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-driver/src/callbacks.rs:216]: Yes, I thought about this -- I'll make this change!

File: flux-driver/src/collector.rs [hunk 2]
```diff
@@ -351,7 +362,7 @@ impl<'tcx, 'a> SpecCollector<'tcx, 'a> {
         }
     }
 
-    fn emit_err<T>(&mut self, err: impl SessionDiagnostic<'a>) -> Result<T, ErrorGuaranteed> {
+    pub fn emit_err<T>(&mut self, err: impl SessionDiagnostic<'a>) -> Result<T, ErrorGuaranteed> {
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-driver/src/collector.rs:365]: hmm good point? I think there was some scope error without it but let me check.

File: flux-typeck/src/wf.rs [hunk 3]
```diff
@@ -302,6 +304,30 @@ impl<'a, 'tcx> Wf<'a, 'tcx> {
...
+                    let e_t = self.synth_expr(env, e)?;
+                    if e_t != *t {
+                        return self.emit_err(errors::SortMismatch::new(e.span, t.clone(), e_t));
+                    }
+                }
+                return Ok(uf_def.output.clone());
+            }
+            return self.emit_err(errors::ParamCountMismatch::new(span, expected, found));
+        }
+        self.emit_err(errors::UnresolvedFunction::new(f.span))
```
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-typeck/src/wf.rs:329]: much nicer, thanks!

File: flux-driver/src/collector.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ranjitjhala) [reply flux-driver/src/collector.rs:365]: fixed!

**Pull request merged**